### ISSUE TITLE
Add RetroShell as a serial port device

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#820]
In another repo, a user requested to control the emulator from within the Amiga to implement a smart warp boot mechanism (switching off warp mode when a specific task has finished).

A clean and easy way to solve this problem is to allow users to (virtually) plug RetroShell into the serial port.

At the moment, there are two serial port devices:

- **Nullmodem cable**
  To play multiplayer games
- **Loopback cable**
  To run diagnostics software

Proposal: Add two more devices:

- **RetroShell viewer**
  This device will listen to the serial port and display the transmitted, readable text in the RetroShell console. Such functionality did already exist. However, it was only possible to enable it via a RetroShell command, and, more importantly, it no longer works in the latest release.

- **RetroShell commander**
  This device will listen to the serial port and translate every readable character into a RetroShell key press in the background. Thus, the transmitted text will be executed as a shell command after '\n' is transmitted. 

I've already hacked together a small prototype as a proof-of-concept implementation:

<img width="872" alt="Bildschirmfoto 2024-03-05 um 08 20 13" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/2f34efbf-9a1e-4824-957c-2532f0579955">

**Comment by dirkwhoffmann** [comment]
Another possibility would be to merge the RetroShell viewer and the RetroShell commander into a single device and to use a unique character to enter command mode.

As this is all about retro, we could use the ASCII character 16 (DLE, Data Link Escape). Once this character is detected, transmitted bytes are no longer just displayed in RetroShell but typed in. Command mode will be reset to text mode on the following line.

**Comment by dirkwhoffmann** [comment]
Unfortunately, I didn't find a way to echo the DLE character in Amiga DOS. There seems to be no way to specify an ASCII character by number inside a string, which would make the special character approach tedious to use. 

If some Amiga DOS expert knows a solution, please let me know.

**Comment by wepl** [comment]
I think there is no way to create a DLE char with the standard tools.
But it would be easy to create a command which outputs it or prepend the output with a DLE.
On 1.3 you anyway need the Echo command to print something.

**Comment by wepl** [comment]
[dle.zip](https://github.com/dirkwhoffmann/vAmiga/files/14508869/dle.zip)
attached a dle command prints the argument prepended with a DLE

**Comment by dirkwhoffmann** [comment]
Thinking more deeply about the available options, I think the dual-device approach is superior. That is because I want to let the Commander send back the RetroShell output to the Amiga eventually, which the viewer (sniffer) should never do. In particular, the two devices will behave as follows:

- **Sniffer**
  Listens to the serial port and displays all incoming and outgoing bytes in RetroShell. This is the same as the Port inspector panel does in the Mac app, just inside RetroShell. 

- **Commander**
  The commander types all outgoing bytes into RetroShell and executes the command when `\n` is received. Eventually, it will send the RetroShell output back to the Amiga through the serial port (I will have to expand the UART class accordingly). 

The important part of the Commander is already implemented (but still needs to be ported over to vAmigaWeb).

In the Mac app, plug in the Commander into the serial port:

<img width="872" alt="Bildschirmfoto 2024-03-06 um 12 40 01" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/1a70e1f5-37c9-4694-8f99-3833b697eb30">

In Kick 3.0 (which allows `ser:` to be addressed directly), send a RetroShell command simply via `echo` (no special characters needed):

<img width="872" alt="Bildschirmfoto 2024-03-06 um 12 40 32" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/8ca671b9-d17a-4399-927b-ffcdd69c95aa">

With the Commander plugged in, the command is executed inside RetroShell: 

<img width="872" alt="Bildschirmfoto 2024-03-06 um 12 40 35" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/b19a2df4-3b89-45d1-9272-726d58a9daef">

With the Commander plugged in, you have complete control over the emulator. You can enable or disable warp mode as you like or anything else.

**Comment by dirkwhoffmann** [comment]
Sending RetroShell output to the Amiga also works (in the lower window, I've typed in `type ser:`) :

<img width="872" alt="Bildschirmfoto 2024-03-06 um 15 06 15" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/ee5e97d0-5387-4054-b2e1-7c4db380f06f">

There are two issues left:
- The known buffering problem. The Amiga buffers incoming data. We need to find out if a flush signal can be sent. Currently, I am not aware of anything like that. 
- More severely, I've experienced two Gurus while testing. My current code might be able to bring the UART to an inconsistent state. More investigation is needed.

**Comment by dirkwhoffmann** [comment]
With the latest code changes, RetroShell can be used inside a Terminal program such as A-Talk III (unlike the `type ser:` solution, there are no buffering issues, and communication is possible with high baud rates).

Important:
- Set handshake to none.
- Select half-duplex mode.
- Press Ctrl+Return to send a command to RetroShell.

<img width="872" alt="Bildschirmfoto 2024-03-07 um 07 16 16" src="https://github.com/dirkwhoffmann/vAmiga/assets/12561945/8089496b-92fe-4123-a591-817744062cab">

There are some caveats, though: 
- Some commands, such as `amiga power off` will crash the emulator. That is because RetroShell commands are expected to be executed within the GUI thread. The RetroShell commander executes the commands within the emulator thread which causes issues with several commands such as all commands that trigger a state change. The emulator will run into an assertion checking the thread type. This will not change until vAmiga 3.0, which will utilize the new code architecture currently being developed for VirtualC64. 

- Terminal emulation is not perfect. For example, typing a backspace in A-Talk III will not result in a backspace inside RetroShell. As a result, correcting typos when entering commands is impossible. Furthermore, RetroShell's clear command does not clear the terminal. 

Summing up:
- The main goal has already been achieved: Providing the means to turn warp mode on and off from within the emulator.
- Everything else is more of a fun project we can work on later during some dull, rainy days.

**Comment by dirkwhoffmann** [comment]
Update: With the latest code changes, the Commander no longer crashes the emulator. 

Before hitting Ctrl+Return:

<img width="1019" alt="Bildschirmfoto 2024-08-06 um 16 37 30" src="https://github.com/user-attachments/assets/e290e25b-b1f2-46bb-8e4c-e63ba26c9ef6">

After hitting Ctrl+Return:
<img width="1019" alt="Bildschirmfoto 2024-08-06 um 16 37 39" src="https://github.com/user-attachments/assets/2c88758d-a6b3-445d-ba80-941d4388ddf4">
